### PR TITLE
Added TQC

### LINKED ISSUE #258: Add TQC to CleanRL
## Problem Description
Would you be interested in adding Truncated Quantile Critics to CleanRL? If so I can work on a PR.

If you are interested, then I have a few questions:
1. Is it okay if I use the new gym.Env api where `env.step` returns `obs, reward, terminated, truncated, info` instead of `obs, reward, done, info`?
2. Is it okay if I deviate slightly from the style followed by the other algorithms in CleanRL? I would like to put the code for calculating the losses and updating the models into a separate class. Everything else will remain the same. This way, if people are familiar with the general RL procedure, and only want to see how the core of the algorithm changes, they will only need to look at this class.

## Checklist
- [x] I have installed dependencies via `poetry install` (see [CleanRL's installation guideline](https://docs.cleanrl.dev/get-started/installation/).
- [x] I have checked that there is no similar [issue](https://github.com/vwxyzjn/cleanrl/issues) in the repo (required)

## Current Behavior
Algorithm is not present

## Expected Behavior
Algorithm will be present

## Steps to Reproduce
1. Look for list of supported algorithms in CleanRL.
3. Find that TQC is not present.

### PR DESCRIPTION
**Pull request by AdityaGudimella**
## Description
Closes #258. Implement [Truncated Quantile Critics](https://paperswithcode.com/paper/controlling-overestimation-bias-with)

## Types of changes

- [ ] Bug fix
- [ ] New feature
- [x] New algorithm
- [ ] Documentation

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/vwxyzjn/cleanrl/blob/master/CONTRIBUTING.md) guide (**required**).
- [x] I have ensured `pre-commit run --all-files` passes (**required**).
- [ ] I have updated the documentation and previewed the changes via `mkdocs serve`.
- [ ] I have updated the tests accordingly (if applicable).

If you are adding new algorithms or your change could result in performance difference, you may need to (re-)run tracked experiments. See https://github.com/vwxyzjn/cleanrl/pull/137 as an example PR. 
- [ ] I have contacted [vwxyzjn](https://github.com/vwxyzjn) to obtain access to the [openrlbenchmark W&B team](https://wandb.ai/openrlbenchmark) (**required**).
- [ ] I have tracked applicable experiments in [openrlbenchmark/cleanrl](https://wandb.ai/openrlbenchmark/cleanrl) with `--capture-video` flag toggled on (**required**).
- [ ] I have added additional documentation and previewed the changes via `mkdocs serve`.
    - [ ] I have explained note-worthy implementation details.
    - [ ] I have explained the logged metrics.
    - [ ] I have added links to the original paper and related papers (if applicable).
    - [ ] I have added links to the PR related to the algorithm.
    - [ ] I have created a table comparing my results against those from reputable sources (i.e., the original paper or other reference implementation).
    - [ ] I have added the learning curves (in PNG format with `width=500` and `height=300`).
    - [ ] I have added links to the tracked experiments.
    - [ ] I have updated the overview sections at the [docs](https://docs.cleanrl.dev/rl-algorithms/overview/) and the [repo](https://github.com/vwxyzjn/cleanrl#overview)
- [ ] I have updated the tests accordingly (if applicable).

### CODE REVIEW

File: cleanrl/tqc_continuous_action.py [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+        help="if toggled, cuda will be enabled by default")
+    parser.add_argument("--track", type=lambda x: bool(strtobool(x)), default=False, nargs="?", const=True,
+        help="if toggled, this experiment will be tracked with Weights and Biases")
+    parser.add_argument("--wandb-project-name", type=str, default="cleanRL",
+        help="the wandb's project name")
+    parser.add_argument("--wandb-entity", type=str, default=None,
+        help="the entity (team) of wandb's project")
+    parser.add_argument("--capture-video",
+        type=lambda x: bool(strtobool(x)), default=False, nargs="?", const=True,
+        help="weather to capture videos of the agent performances (check out `videos` folder)")
```
Comment (vwxyzjn) [inline cleanrl/tqc_continuous_action.py:53]: Style nit: please keep this part consistent with other scripts. The first line is the argument, and the second line is the help text / documentation

File: cleanrl/tqc_continuous_action.py [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+        help="total timesteps of the experiments")
+    parser.add_argument("--buffer-size", type=int, default=int(1e6),
+        help="the replay memory buffer size")
+    parser.add_argument("--gamma", type=float, default=0.99,
+        help="the discount factor gamma")
+    parser.add_argument("--tau", type=float, default=0.005,
+        help="target smoothing coefficient (default: 0.005)")
+    parser.add_argument("--batch-size", type=int, default=256,
+        help="the batch size of sample from the reply memory")
+    parser.add_argument("--handle-timeout-termination",
```
Comment (vwxyzjn) [inline cleanrl/tqc_continuous_action.py:68]: Style nit: the same comment as above.

File: cleanrl/tqc_continuous_action.py [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+        type=lambda x: bool(strtobool(x)), default=False,
+        help="treat TimeLimit.truncated == True as done == False")
+    parser.add_argument("--n-quantiles", type=int, default=25,
+        help="the number of quantiles used for each Q Network")
+    parser.add_argument("--n-critics", type=int, default=5,
+        help="the number of Q Networks")
+    parser.add_argument("--use-paper-n-quantiles-to-drop",
```
Comment (vwxyzjn) [inline cleanrl/tqc_continuous_action.py:75]: Style nit: the same comment as above.

File: cleanrl/tqc_continuous_action.py [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+        help="the learning rate to tune target entropy")
+    args = parser.parse_args()
+    # fmt: on
+    return args
+
+
+def layer_init(layer, std=np.sqrt(2), bias_const=0.0):
+    torch.nn.init.orthogonal_(layer.weight, std)
+    torch.nn.init.constant_(layer.bias, bias_const)
+    return layer
```
Comment (vwxyzjn) [inline cleanrl/tqc_continuous_action.py:97]: Does the original paper use PPO-style initialization? If the paper doesn't use it, we should not use it by default.  The implementations at CleanRL should try to match the original implementations whenever possible.

File: cleanrl/tqc_continuous_action.py [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+    torch.manual_seed(args.seed)
+    torch.backends.cudnn.deterministic = args.torch_deterministic
+
+    device = torch.device("cuda" if torch.cuda.is_available() and args.cuda else "cpu")
+
+    # env setup
+    envs = gym.vector.SyncVectorEnv([make_env(args.env_id, args.seed, 0, args.capture_video, run_name)])
+    assert isinstance(envs.single_action_space, gym.spaces.Box), "only continuous action space is supported"
+
+    max_action = float(envs.single_action_space.high[0])
```
Comment (vwxyzjn) [inline cleanrl/tqc_continuous_action.py:237]: This variable is not used. Hmmm pre-commit should have caught this 🤔

File: cleanrl/tqc_continuous_action.py [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+    start_time = time.time()
+
+    # TRY NOT TO MODIFY: start the game
+    obs = envs.reset()
+    for global_step in range(args.total_timesteps):
+        # ALGO LOGIC: put action logic here
+        if global_step < args.learning_starts:
+            actions = np.array([envs.single_action_space.sample() for _ in range(envs.num_envs)])
+        else:
+            actions, _, _ = actor.get_action(torch.Tensor(obs).to(device))
```
Comment (vwxyzjn) [inline cleanrl/tqc_continuous_action.py:274]: I suggest enclosing this block with `with torch.no_grad():`

**Review by vwxyzjn — COMMENTED** [review]
Looks great! Had a preliminary pass. I suggest at least running a few experiments to see if the performance matches the reference implementation / paper's results. After we are done several passes, I will ask Antonin from SB3, who has implemented TQC, to see if he is interested to help us review :)

File: cleanrl/tqc_continuous_action.py (hunk 1 above)
Comment (AdityaGudimella) [reply cleanrl/tqc_continuous_action.py]: Updated

File: cleanrl/tqc_continuous_action.py (hunk 2 above)
Comment (AdityaGudimella) [reply cleanrl/tqc_continuous_action.py]: Updated

File: cleanrl/tqc_continuous_action.py (hunk 3 above)
Comment (AdityaGudimella) [reply cleanrl/tqc_continuous_action.py]: Updated

File: cleanrl/tqc_continuous_action.py (hunk 4 above)
Comment (AdityaGudimella) [reply cleanrl/tqc_continuous_action.py]: Just checked the author's implementation. The paper does not use any initialization.

File: cleanrl/tqc_continuous_action.py (hunk 5 above)
Comment (AdityaGudimella) [reply cleanrl/tqc_continuous_action.py]: Removed unused variable.

File: cleanrl/tqc_continuous_action.py (hunk 6 above)
Comment (AdityaGudimella) [reply cleanrl/tqc_continuous_action.py]: Done.

**Comment by AdityaGudimella** [comment]
I can run the algo against the same mujoco envs as run in the paper. Would once against each env be sufficient? Also how do I share the results of the run with you?

**Comment by vwxyzjn** [comment]
Thank you @AdityaGudimella. The variant looks good. I suggest running some preliminary experiments in your own wandb namespace and create a wandb report to share the findings.

**Comment by AdityaGudimella** [comment]
Apologies for the delay in this. I ran 2 trails each of Hopper-v3, Humanoid-v3 and Swimmer-v3 and didn't have any resources available to run experiments after that. I've just set 2 trials each of HalfCheetah-v3, Ant-v3 and Walker2d-v3 now. Once those experiments are done (probably in 2 days) I will share the report here.

**Comment by vwxyzjn** [comment]
That sounds good. Feel free to ping me if this is ready for review or if there is anything I can help.
